File: General-DSA-Questions/Some Named Algo & Genral Ques/6_3Sum.cpp
// The 3 Sum problem asks you to find all unique triplets in an array that sum up to a given target value (usually zero).


// Approach and Intuition:(Fix one element and use two pointer approach for the other two elements)
// 1. Sorting:
//    a.  First, sort the array to help easily skip duplicate elements and apply a two-pointer approach efficiently.
// 2. Iterate with a fixed element:
//    a.  The idea is to fix one element of the triplet (nums[i]), and then find the remaining two numbers that sum up
//        to -nums[i] (because we are aiming for the sum to be zero). We do this using the two-pointer technique.
// 3. Two-pointer technique:
//    a.  After fixing one element, use two pointers (left and right) to find pairs whose sum is equal to -nums[i].
//        The pointers start at the next index (left = i + 1) and the end of the array (right = n - 1).
//        a1. If the sum of nums[i] + nums[left] + nums[right] is less than zero, we need a larger sum,
//            so we move the left pointer to the right (left++).
//        a2. If the sum is greater than zero, we need a smaller sum, so we move the right pointer to the left (right--).
//        a3. If the sum is exactly zero, we record the triplet and move both pointers inward to avoid duplicates.
// 4. Avoid duplicates:
//    a.  After finding a valid triplet, skip over any duplicate values for nums[i], nums[left], and nums[right] to ensure uniqueness.


#include <iostream>
#include <vector>
#include <algorithm>  // For sort method
using namespace std;

vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> result;
    int n = nums.size();
    
    // Step 1: Sort the array
    sort(nums.begin(), nums.end());
    
    // Step 2: Iterate through the array
    for (int i = 0; i < n - 2; i++) {   // We need at least 3 elements to form a triplet, so we go up to n-2
        // Skip duplicate elements for the first element of the triplet
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        
        int left = i + 1;
        int right = n - 1;
        
        // Step 3: Use two-pointer technique to find the other two elements
        while (left < right) {
            int sum = nums[i] + nums[left] + nums[right];

            if (sum < 0) {
                left++;  // We need a larger sum
            } else if (sum > 0) {
                right--; // We need a smaller sum
            } else {
                // Found a triplet
                result.push_back({nums[i], nums[left], nums[right]});
                
                // Skip duplicate elements for the second element of the triplet
                while (left < right && nums[left] == nums[left - 1]) left++;
                // Skip duplicate elements for the third element of the triplet
                while (left < right && nums[right] == nums[right + 1]) right--;
                
                // Move both pointers inward
                left++;
                right--;
            }
        }
    }
    return result;
}

int main() {
    vector<int> nums = {-1, 0, 1, 2, -1, -4};
    vector<vector<int>> triplets = threeSum(nums);
    
    cout << "Unique triplets that sum to zero are:\n";
    for (const auto& triplet : triplets) {
        cout << "[";
        for (size_t i = 0; i < triplet.size(); i++) {
            cout << triplet[i];
            if (i < triplet.size() - 1) cout << ", ";
        }
        cout << "]\n";
    }
    
    return 0;
}
// Time Complexity: O(n^2) - Sorting the array takes O(n log n)
// and the two-pointer technique takes O(n) for each of the n elements.
// Space Complexity: O(1) - We are using only a constant amount of extra space (not counting the output storage).

// 4 Sum Problem is also similar to this problem but there we fix two elements and then use two pointer approach for the other two elements.

File: JS-DSA-Questions/Some Named Algo & Genral Ques/6_3Sum.js
// The 3 Sum problem asks you to find all unique triplets in an array that sum up to a given
// target value (usually zero).

// Approach and Intuition: (Fix one element and use two pointers for the other two)
// 1. Sorting: sort the array to easily skip duplicates and apply two pointers.
// 2. Fix nums[i], then find two numbers summing to -nums[i] using two pointers.
// 3. Two-pointer technique with left = i+1 and right = n-1:
//    a. If sum < 0, move left right (need larger sum).
//    b. If sum > 0, move right left (need smaller sum).
//    c. If sum == 0, record the triplet and move both pointers inward.
// 4. Avoid duplicates: skip repeated values for nums[i], nums[left], and nums[right].

// JavaScript note: sort numbers with (a, b) => a - b (default sort is lexicographic).
function threeSum(nums) {
  const result = [];
  const n = nums.length;

  // Step 1: Sort the array
  nums.sort((a, b) => a - b);

  // Step 2: Iterate; we need at least 3 elements, so go up to n-2
  for (let i = 0; i < n - 2; i++) {
    // Skip duplicate first elements
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let left = i + 1;
    let right = n - 1;

    // Step 3: Two-pointer search
    while (left < right) {
      const sum = nums[i] + nums[left] + nums[right];

      if (sum < 0) {
        left++; // need a larger sum
      } else if (sum > 0) {
        right--; // need a smaller sum
      } else {
        // Found a triplet
        result.push([nums[i], nums[left], nums[right]]);

        left++;
        right--;

        // Skip duplicates for the second and third elements
        while (left < right && nums[left] === nums[left - 1]) left++;
        while (left < right && nums[right] === nums[right + 1]) right--;
      }
    }
  }
  return result;
}

console.log(threeSum([-1, 0, 1, 2, -1, -4])); // [[-1,-1,2],[-1,0,1]]

// Time Complexity: O(n^2). Space Complexity: O(1) (not counting output).
// The 4 Sum problem is similar, but there we fix two elements and use two pointers for the rest.
